File: lib/linear/models/comment.rb
# frozen_string_literal: true

require 'gqli'

module Rubyists
  # Namespace for Linear
  module Linear
    M :base_model, :user
    Comment = Class.new(BaseModel)
    # The Comment class represents a Linear issue comment.
    class Comment
      include SemanticLogger::Loggable
      many_to_one :user

      Base = fragment('BaseComment', 'Comment') do
        id
        body
        url
        user { ___ User.base_fragment }
        createdAt
        updatedAt
      end

      def to_s
        format('%<id>-12s %<url>s', id:, url:)
      end

      def inspection
        format('id: "%<id>s" url: "%<url>s"', id:, url:)
      end
    end
  end
end

File: app/lib/linear_cli/linear/comment.ex
defmodule LinearCli.Linear.Comment do
  @moduledoc """
  A Linear issue comment. Ported from vendor/ruby-linear-cli/lib/linear/models/comment.rb.

  Never queried top-level in the Ruby original - only ever nested inside an
  issue's fragment (or created via the `commentCreate` mutation, ported in a
  later phase alongside the other write commands). No actions defined here yet.
  """

  use Ash.Resource, domain: LinearCli.Linear

  attributes do
    attribute :id, :string, primary_key?: true, allow_nil?: false, public?: true
    attribute :body, :string, public?: true
    attribute :url, :string, public?: true
    attribute :user, :term, public?: true
  end

  @doc "GraphQL field selection for a comment's own fields (Ruby: Comment::Base)."
  def base_fields do
    "id body url user { #{LinearCli.Linear.User.fields_with_teams()} } createdAt updatedAt"
  end

  @doc false
  def from_map(map) do
    struct!(__MODULE__,
      id: map["id"],
      body: map["body"],
      url: map["url"],
      user: map["user"] && LinearCli.Linear.User.from_map(map["user"])
    )
  end
end
